### PR TITLE
Fix HEIGHT_SHARDED width front-pad offset

### LINKED ISSUE #49492: `ttnn.pad`: HEIGHT_SHARDED + width front-padding + sharded output produces wrong results
### Description

When `ttnn.pad` is called with a HEIGHT_SHARDED RM input, width front-padding (non-zero `before` on the W axis), and a HEIGHT_SHARDED output, the result is garbage (PCC ~0.002 vs torch reference).

The bug is in the `PadRmShardedWidthOnlyProgramFactory` which miscomputes front-pad offsets. This factory is invoked when `pad.cpp` detects HEIGHT_SHARDED input with width padding and a non-interleaved output config — it enters the HEIGHT_SHARDED-specific width-pad decomposition path.

**Note:** HEIGHT_SHARDED + width front-pad + DRAM (interleaved) output works correctly (PCC 1.0) because it bypasses the sharded factory and uses the default `PadRmReaderWriterMultiCoreDefaultProgramFactory`.

### Reproducer

```python
import torch
import ttnn

device = ttnn.open_device(device_id=0)

input_shape = [1, 1, 64, 128]
padding_spec = [(0, 0), (0, 0), (0, 0), (64, 0)]  # front-pad W by 64

torch.manual_seed(0)
torch_input = torch.randn(input_shape, dtype=torch.bfloat16)
torch_output = torch.nn.functional.pad(torch_input, [64, 0, 0, 0], value=0.0)

# Create HEIGHT_SHARDED input
num_cores = device.compute_with_storage_grid_size()
total_cores = min(num_cores.x * num_cores.y, 64)
shard_h = (64 + total_cores - 1) // total_cores
grid = ttnn.CoreRangeSet([ttnn.CoreRange(
    ttnn.CoreCoord(0, 0),
    ttnn.CoreCoord(min(total_cores, num_cores.x) - 1, (total_cores - 1) // num_cores.x)
)])
in_cfg = ttnn.MemoryConfig(
    ttnn.TensorMemoryLayout.HEIGHT_SHARDED, ttnn.BufferType.L1,
    ttnn.ShardSpec(grid, (shard_h, 128), ttnn.ShardOrientation.ROW_MAJOR)
)
out_cfg = ttnn.MemoryConfig(
    ttnn.TensorMemoryLayout.HEIGHT_SHARDED, ttnn.BufferType.L1,
    ttnn.ShardSpec(grid, (shard_h, 192), ttnn.ShardOrientation.ROW_MAJOR)
)

tt_input = ttnn.from_torch(torch_input, layout=ttnn.ROW_MAJOR_LAYOUT,
                           dtype=ttnn.bfloat16, device=device, memory_config=in_cfg)
tt_output = ttnn.pad(tt_input, padding_spec, value=0.0, memory_config=out_cfg)
actual = ttnn.to_torch(tt_output)

pcc = torch.corrcoef(torch.stack([torch_output.float().flatten(),
                                   actual.float().flatten()]))[0, 1].item()
print(f"PCC: {pcc:.6f}")  # Prints ~0.002250

ttnn.close_device(device)
```

### Expected behavior
PCC should be >= 0.9999.

### Actual behavior
PCC is ~0.002 — output is effectively garbage.

### Root cause
`PadRmShardedWidthOnlyProgramFactory` does not correctly handle the `before` (front-pad) offset on the width dimension. The factory was designed for trailing-only width padding on HEIGHT_SHARDED tensors.

### PR DESCRIPTION
**Pull request by VirdhatchaniKN**
### Issue
https://github.com/tenstorrent/tt-metal/issues/49492

### Summary
Fixes a pre-existing bug where `ttnn.pad` with a **HEIGHT_SHARDED** RM input, **width front-padding**, and **HEIGHT_SHARDED** output produced garbage results (PCC ~0.002 vs torch reference).

The root cause was a one-line indexing error in `PadRmShardedWidthOnlyProgramFactory`: the width front-pad byte offset was read from `input_tensor_start[-3]` (C dimension) instead of `input_tensor_start[3]` (W dimension). The reader kernel then placed input data at offset 0, effectively turning front-padding into back-padding.

### Root cause

For `padding_spec = [(0,0), (0,0), (0,0), (64, 0)]`, `input_tensor_start` is `[0, 0, 0, 64]`:

| Index | Dim | Value |
|-------|-----|-------|
| `[-3]` | C | 0 (wrong — was used) |
| `[3]` | W | 64 (correct) |

With `W_padding_front_bytes = 0`, the reader wrote input sticks starting at byte 0 of each output stick instead of after 64 columns of padding.

**Note:** HEIGHT_SHARDED + width front-pad + **DRAM** (interleaved) output already worked (PCC 1.0) because it bypasses this factory and uses `PadRmReaderWriterMultiCoreDefaultProgramFactory`.

**Fix** is to change `input_tensor_start[-3]` to `input_tensor_start[3]` in `PadRmShardedWidthOnlyProgramFactory` (data_movement and quasar copies).

### Pipeline status

- [x] Sanity
- [x] (Runtime) Sanity Test
- [x] BHPC
- [x] Nightly for all categories
- [x] All model test
- [x] Single card
- [x] T3K
- [x] Galaxy

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:virdhatchani/pad_bug)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:virdhatchani/pad_bug)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-sanity-tests.yaml?query=branch:virdhatchani/pad_bug)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:virdhatchani/pad_bug)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:virdhatchani/pad_bug)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:virdhatchani/pad_bug)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:virdhatchani/pad_bug)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:virdhatchani/pad_bug)

### CODE REVIEW

File: tests/ttnn/nightly/unit_tests/operations/data_movement/test_pad_universal_input.py [hunk 1]
```diff
@@ -381,7 +380,46 @@ def test_pad_rm_wb_sharded_front_pad_to_sharded(
...
 
 # ---------------------------------------------------------------------------
-# Category 7: DRAM WIDTH/BLOCK sharded input (to_memory_config composite fallback)
+# Category 7: RM HEIGHT_SHARDED — width front-padding
+#
+# front_pad_w duplicates FRONT_PAD_CASES shapes but uses HEIGHT sharding, so it
+# routes through PadRmShardedWidthOnlyProgramFactory instead of the W/B RM path.
+# ---------------------------------------------------------------------------
+
+HEIGHT_FRONT_PAD_CASES = [
```
Comment (Aswinmcw) [inline tests/ttnn/nightly/unit_tests/operations/data_movement/test_pad_universal_input.py:389]: Optional (non-blocking): these HEIGHT_SHARDED front-pad cases only exercise bf16. The line this PR fixes computes `W_padding_front_bytes = input_tensor_start[3] * element_size`, so adding a float32 case here (element_size=4) would additionally guard the byte-offset scaling on the exact changed path — the existing float32 coverage (Cat 9) runs through the W/B factory, not `PadRmShardedWidthOnlyProgramFactory`.

**Review by Aswinmcw — APPROVED** [review]
LGTM. Correct root cause and minimal fix — verified the reader kernel copies input at `output_stick_ptr[W_front_pad_bytes + i]`, so sourcing the offset from W (`[3]`) rather than C (`[-3]`) is what makes front-pad actually front-pad, and the quasar copy is kept in sync.

**Review by mateusznowakTT — APPROVED** [review]

**Review by bbradelTT — APPROVED** [review]

**Review by amorrisonTT — APPROVED** [review]
